packages/primitives/popover/src/popover-root.directive.ts

```diff
@@ -142,20 +158,97 @@ export class RdxPopoverRootDirective implements OnInit {
+        };
+    }
+
+    /** @ignore */
+    private onStateChangeEffect() {
         effect(() => {
-            const isOpen = this.isOpen();
+            const state = this.state();
 
             untracked(() => {
-                if (isOpen) {
-                    this.show();
-                } else {
-                    this.hide();
+                isRdxPopoverDevMode() &&
+                    console.log(
+                        this.uniqueId(),
+                        '[onStateChangeEffect] pre filter',
+                        state,
+                        this.getAnimationParamsSnapshot()
+                    );
+                if (!this.ifRunOnStateChangeCallbackSync(state)) {
+                    return;
                 }
+                isRdxPopoverDevMode() && console.log(this.uniqueId(), '[onStateChangeEffect]', state);
+                this.onStateChangeCallback(state);
             });
         });
     }
 
+    /** @ignore */
+    private onCssAnimationStatusChangeChangeEffect() {
+        effect(() => {
+            const cssAnimationStatus = this.cssAnimationStatus();
+
+            untracked(() => {
+                isRdxPopoverDevMode() &&
+                    console.log(
+                        this.uniqueId(),
+                        '[onCssAnimationStatusChangeChangeEffect pre filter]',
+                        cssAnimationStatus,
+                        this.getAnimationParamsSnapshot()
+                    );
+
+                if (!this.ifRunOnStateChangeCallbackAsync(cssAnimationStatus)) {
+                    return;
+                }
+
+                isRdxPopoverDevMode() &&
+                    console.log(
+                        this.uniqueId(),
+                        '[onCssAnimationStatusChangeChangeEffect]',
+                        cssAnimationStatus,
+                        this.state()
+                    );
+                this.onStateChangeCallback(this.state());
+            });
+        });
+    }
+
+    /** @ignore */
+    private ifRunOnStateChangeCallbackSync(state: RdxPopoverState) {
+        return (
+            !this.popoverContentAttributesDirective() ||
+            !this.cssAnimation() ||
+            !this.cssOnCloseAnimation() ||
+            state === RdxPopoverState.OPEN
+        );
+    }
+
+    /** @ignore */
+    private ifRunOnStateChangeCallbackAsync(cssAnimationStatus: RdxPopoverAnimationStatus | null) {
+        return (
+            this.popoverContentAttributesDirective() &&
+            this.cssAnimation() &&
+            this.cssOnCloseAnimation() &&
+            cssAnimationStatus &&
+            this.state() === RdxPopoverState.CLOSED &&
+            [RdxPopoverAnimationStatus.CLOSED_ENDED].includes(cssAnimationStatus)
+        );
+    }
+
+    /** @ignore */
+    private onStateChangeCallback(state: RdxPopoverState) {
+        const isOpen = this.isOpen(state);
+        isOpen ? this.show() : this.hide();
+        this.onOpenChange.emit(isOpen);
```

Comment:
To be fixed: `open` event should get emitted once the open animation is done instead of the current synchronous emission

Reply:
fixed